[DevTools] Show component names while highlighting renders

piotrski opened the pull request:
## Summary
This PR improves the Trace Updates feature by letting developers see component names directly on the update overlay. Before this change, the overlay only highlighted updated regions, leaving it unclear which components were involved. With this update, you can now match visual updates to their corresponding components, making it much easier to debug rendering performance.

### New Feature: Show component names while highlighting
When the new **"Show component names while highlighting"** setting is enabled, the update overlay display the names of affected components above the rectangles, along with the update count. This gives immediate context about what’s rendering and why. The preference is stored in local storage and synced with the backend, so it’s remembered across sessions.

### Improvements to Drawing Logic
The drawing logic has been updated to make the overlay sharper and easier to read. Overlay now respect device pixel ratios, so they look great on high-DPI screens. Outlines have also been made crisper, which makes it easier to spot exactly where updates are happening.

> [!NOTE]
> **Grouping Logic and Limitations**
> Updates are grouped by their screen position `(left, top coordinates)` to combine overlapping or nearby regions into a single group. Groups are sorted by the highest update count within each group, making the most frequently updated components stand out.
> Overlapping labels may still occur when multiple updates involve components that overlap but are not in the exact same position. This is intentional, as the logic aims to maintain a straightforward mapping between update regions and component names without introducing unnecessary complexity.

### Testing
This PR also adds tests for the new `groupAndSortNodes` utility, which handles the logic for grouping and sorting updates. The tests ensure the behavior is reliable across different scenarios.

## Before & After

https://github.com/user-attachments/assets/6ea0fe3e-9354-44fa-95f3-9a867554f74c

https://github.com/user-attachments/assets/32af4d98-92a5-47dd-a732-f05c2293e41b

piotrski committed:
refactor(trace-updates): simplify `iterateNodes` function and improve type consistency

piotrski committed:
feat(trace-updates): group overlapping updates and merge labels

piotrski committed:
feat(trace-updates): display update count on component labels

piotrski committed:
feat(trace-updates): sort component update labels by maximum update count

piotrski committed:
fix(trace-updates): add component name display to canvas

piotrski committed:
feat(trace-updates): add setting to show component names during trace highlighting

piotrski committed:
fix(trace-updates): improve canvas rendering on high-DPI screens

piotrski committed:
refactor(trace-updates): improve draw logic and modularize canvas rendering

piotrski committed:
test(trace-updates): add tests for node grouping and sorting logic

hoxyq commented on packages/react-devtools-shared/src/backend/views/TraceUpdates/canvas.js:
```diff
@@ -34,7 +32,7 @@ let canvas: HTMLCanvasElement | null = null;
 function drawNative(nodeToData: Map<HostInstance, Data>, agent: Agent) {
   const nodesToDraw = [];
-  iterateNodes(nodeToData, (_, color, node) => {
+  iterateNodes(nodeToData, ({color, node}) => {
     nodesToDraw.push({node, color});
   });
```
For React Native, we should probably add new fields here as well, including grouping via `groupAndSortNodes`.

For backwards compatibility, lets add a new event for it, something like `drawGroupedTraceUpdatesWithNames`. 

```
...

agent.emit('drawTraceUpdates', nodesToDraw);
agent.emit('drawGroupedTraceUpdatesWithNames', mergedNodes);
```

hoxyq commented:
Love these improvements, thanks for opening this PR! Left some comments. I will try to get this supported for React Native, and I am also thinking about moving the checkbox to some other place, so this feature gets more visibility.

piotrski commented on packages/react-devtools-shared/src/devtools/views/Settings/GeneralSettings.js:
```diff
@@ -83,6 +85,19 @@ export default function GeneralSettings(_: {}): React.Node {
             />{' '}
             Highlight updates when components render.
           </label>
+          <div className={styles.Setting}>
+            <label>
+              <input
+                type="checkbox"
+                checked={showNamesWhenTracing}
+                disabled={!traceUpdatesEnabled}
+                onChange={({currentTarget}) =>
+                  setShowNamesWhenTracing(currentTarget.checked)
+                }
+              />{' '}
+              Show component names while highlighting.
+            </label>
+          </div>
```
I left it visible with a disabled style for now to help with discoverability. I also have a separate branch where "Highlight updates when components render." is a toggle in the navbar next to the search, but I didn’t want to make this PR bigger, and I wasn’t sure which icons to use. Open to suggestions!
 
<img width="407" alt="Screenshot 2024-12-03 at 14 39 27" src="https://github.com/user-attachments/assets/e2a597a9-57dc-464e-be90-c2df59ce7c59">

hoxyq replied:
> I left it visible with a disabled style for now to help with discoverability.

Nice, this should be enough for now

> I also have a separate branch where "Highlight updates when components render." is a toggle in the navbar next to the search

Feel free to open a separate PR for this, or the one on top of that

piotrski commented on packages/react-devtools-shared/src/backend/views/TraceUpdates/canvas.js:
```diff
@@ -46,62 +44,132 @@ function drawWeb(nodeToData: Map<HostInstance, Data>) {
+    const colorIndex = Math.min(COLORS.length - 1, data.count - 1);
     const color = COLORS[colorIndex];
-    execute(rect, color, node);
+    execute({...data, color, node});
```
I can convert it back to attributes if needed. I went with an object here because it felt more type-safe and flexible, but let me know what you think.

hoxyq replied:
> I can convert it back to attributes if needed. I went with an object here because it felt more type-safe and flexible, but let me know what you think.

This might become expensive, if the whole page re-renders? For each HTMLElement you will create a temporary copy here just for iteration purposes.

piotrski replied:
Sorry for the delay, I’ve been unwell. I replaced the spread operator with explicit object construction to avoid extra copies. Let me know if this solves the issue for you.

piotrski commented on packages/react-devtools-shared/src/backend/views/TraceUpdates/index.js:
```diff
@@ -86,6 +96,11 @@ function traceUpdates(nodes: Set<HostInstance>): void {
       rect = measureNode(node);
     }
 
+    let displayName = agent.getComponentNameForHostInstance(node);
```
Good point. Do you think there’s a cleaner way to check if the component was optimized by the compiler? I saw the `parseElementDisplayNameFromBackend` utility can return `compiledWithForget`, but it needs the`type`. Not sure if I can get that from `HostNode` since it’s typed as an object.

hoxyq replied:
> Not sure if I can get that from HostNode since it’s typed as an object.

Here we operate with `HostInstance`: `HTMLElement` for Web (with `react-dom`), and something completely different for Native, based on version and architecture of React Native. 

You could leave it like this for now, maybe later we could add a new method to `Agent`, which would map `HostInstance` to `DevToolsInstance` (this is what `getComponentNameForHostInstance` is already doing), and then return if component is compiled or not, and maybe filter out HOC names.

piotrski committed:
fix(trace-updates): conditionally get component names based on setting

piotrski committed:
refactor(trace-updates): replace spread operator with explicit object construction

piotrski committed:
fix: remove console log

piotrski committed:
feat(trace-updates): add grouped trace updates with names for React Native

- Introduced the `drawGroupedTraceUpdatesWithNames` event in `Agent` for emitting grouped trace updates.
- This maintains backward compatibility while improving trace update organization for React Native.

Note: This change only covers the JavaScript implementation; the native side still needs to be implemented.

hoxyq commented on packages/react-devtools-shared/src/backend/views/TraceUpdates/index.js:
```diff
@@ -86,6 +96,11 @@ function traceUpdates(nodes: Set<HostInstance>): void {
       rect = measureNode(node);
     }
 
+    let displayName = this.showNames && agent.getComponentNameForHostInstance(node);
+    if (displayName != null && displayName.startsWith('Forget(')) {
+      displayName = '✨' + displayName.slice(7, -1);
+    }
```
There is a chance you might end up receiving `TypeError`: calling `displayName.startsWith(...)` if `this.showNames` is `false`.

piotrski replied:
Thank for pointing that out. I’ve updated the code to avoid the potential TypeError by ensuring `displayName` is only processed when `showNames` is true.

I also revamped the display name parsing logic to make it more extensible. As part of this, I added a 🧠 icon for components wrapped in `React.Memo`, and display names are now consistently unwrapped from HOCs for better clarity. Let me know if there’s anything else!

hoxyq reviewed (approved):
Left 1 comment, please address it, linter and Flow issues.

Looks good overall, so once this is fixed, I can merge this

piotrski committed:
feat(trace-updates): extract HOC names and add markers to display names

piotrski committed:
fix: add missing return type

hoxyq committed:
fix[utils]: lint

hoxyq committed:
fix[TraceUpdates]: update Forget marker prefix

AlexVidoso commented:
It's a nice feature, but something went wrong. I can't disable the display of component names, and in some cases, this feature is quite annoying. Could you please check? I'm using Thorium 128 on Linux and tried to disable/enable/reinstall the extension
![image](https://github.com/user-attachments/assets/eb6b40ba-d4f1-4f7d-a424-1ac8f9aa76cb)

hoxyq commented:
>  I can't disable the display of component names

Component names will be displayed by default if this feature is enabled. You can disable this feature by unchecking "Highlight updates when component render." checkbox. Does that work?

hoxyq commented:
> there is no checkbox for "Show component names..."

This is expected. It was removed in https://github.com/facebook/react/pull/32019

AlexVidoso commented:
In my opinion, that's a significant oversight. Nevertheless, thank you for the information.

vision-jinwoo commented:
@hoxyq I agree with @AlexVidoso.
I don't think we always need a rendering name. I'd like to see it as optional. ( when simply debugging, there is no need to know the rendering name. )

alanloi commented:
Agree with the comments from @vision-jinwoo and @AlexVidoso. In some situations when there is a lot going on on the screen, the component names can get in the way of seeing the boxes. It would be better to have the flexibility of disabling the names whilst still being able to see the box highlighting, like in the original PR from @piotrski. As it is right now, it makes this feature much less usable than before the change.

jayers-scg commented:
Adding to voices above: I have started turning this feature off where I used to leave it on all the time, because sometimes components are grouped in such a way or update so often that the screen becomes significantly cluttered and visually annoying while I'm trying to work. This feature is helpful but I would love if I could turn it off -- I could leave the render highlights on always while working in the past without it cluttering my workspace.

As it stands now I'm going to start turning it off until I'm explicitly testing performance, which is unfortunate considering it would sometimes help me find performance issues when I wasn't looking for them in the past.

4anghyeon commented:
I too am sad that the checkbox option is gone. so I inserted the script below to turn off this feature.

```js
window.__REACT_DEVTOOLS_GLOBAL_HOOK__.reactDevtoolsAgent.getComponentNameForHostInstance = function() { return null; };
```

And I created a very simple [chrome extension ](https://github.com/4anghyeon/react-dev-tools-turn-off-component-name)for this for my own use.

I hope that one day the "Show component names" checkbox to come back.

alanloi commented:
> I too am sad that the checkbox option is gone. so I inserted the script below to turn off this feature.
> 
> ```js
> window.__REACT_DEVTOOLS_GLOBAL_HOOK__.reactDevtoolsAgent.getComponentNameForHostInstance = function() { return null; };
> ```
> 
> And I created a very simple [chrome extension ](https://github.com/4anghyeon/react-dev-tools-turn-off-component-name)for this for my own use.
> 
> I hope that one day the "Show component names" checkbox to come back.

That works perfectly, thanks for sharing this! Workarounds like this really shouldn't be needed if the feature was implemented with better user flexibility, but hey here we are..
